### PR TITLE
[RELAY] Refactor FoldConstant to skip TNonComputationalOps

### PR DESCRIPTION
**Pull request by electriclilies**
This PR refactors FoldConstant to skip TNonComputational ops, instead of explicitly checking for those ops. It also makes all QNN ops TNonComputational. This allows the constant folder to be applied to graphs containing QNN ops. 

@jroesch @jwfromm please take a look!

### CODE REVIEW

File: src/relay/qnn/op/concatenate.cc [hunk 1]
```diff
@@ -207,6 +207,7 @@ RELAY_REGISTER_OP("qnn.concatenate")
                   "The quantization zero_point of the output tensor.")
     .set_support_level(11)
     .add_type_rel("QnnConcatenate", QnnConcatenateRel)
+    .set_attr<TNonComputational>("TNonComputational", true)
```
Comment (jwfromm) [inline src/relay/qnn/op/concatenate.cc:210]: Where is `TNonComputational` defined? I don't see it in this PR.

File: src/relay/transforms/fold_constant.cc [hunk 2]
```diff
@@ -151,9 +151,12 @@ class ConstantFolder : public MixedModeMutator {
     }
 
     // We should think about potentially constant evaluation over these ops too.
-    if (call->op == invoke_tvm_op_ || call->op == shape_func_op_ || call->op == alloc_tensor_op_ ||
-        call->op == alloc_storage_op_ || call->op == device_copy_op_) {
```
Comment (jwfromm) [inline src/relay/transforms/fold_constant.cc:155]: Do we need to add all these ops as non-computational? It looks like with this change things like `shape_func_op_` will now be handled differently.

File: src/relay/qnn/op/concatenate.cc (hunk 1 above)
Comment (electriclilies) [reply src/relay/qnn/op/concatenate.cc:210]: TNonComputational is defined in include/tvm/relay/op_attr_types.h

File: src/relay/transforms/fold_constant.cc (hunk 2 above)
Comment (electriclilies) [reply src/relay/transforms/fold_constant.cc:155]: These ops are already marked as TNonComputational when they are registered

**Comment by jwfromm** [comment]
LGTM, thanks!

**Review by jwfromm — APPROVED** [review]

**Review by jroesch — APPROVED** [review]

File: src/relay/transforms/fold_constant.cc (hunk 2 above)
Comment (electriclilies) [reply src/relay/transforms/fold_constant.cc:155]: I missed device_copy initially. It seems that marking device_copy as TNonComputational causes some problems in the fuse_ops pass, so I am leaving the explicit check for device_copy

File: src/relay/transforms/fold_constant.cc [hunk 3]
```diff
@@ -151,9 +151,12 @@ class ConstantFolder : public MixedModeMutator {
...
-      return GetRef<Call>(call);
+    static auto fnoncomputational = Op::GetAttrMap<TNonComputational>("TNonComputational");
+    if (auto call_node = call->op.as<OpNode>()) {
```
Comment (junrushao1994) [inline src/relay/transforms/fold_constant.cc:155]: Let’s be more specific about the type although they are the same
```suggestion
    if (const auto* call_node = call->op.as<OpNode>()) {
```

**Comment by jwfromm** [comment]
@junrushao1994 now that this is passing CI, can you take a final look and merge this?

**Comment by junrushao1994** [comment]
Sorry for being late. It looks good! Thanks!

**Pull request merged**